### PR TITLE
main.coffee: Check for existence of .coafile

### PR DESCRIPTION
**Pull request by EverWinter23**
Fixes https://github.com/coala/coala-atom/issues/60

### CODE REVIEW

**Comment by satwikkansal** [comment]
```
Also resolves following atom dependecy issue:
Package libnotify64 not installed
```
This should be separate commit imo

**Comment by satwikkansal** [comment]
Also it'd be better to create a proper issues for the "NLTK data" and "libnotify64", and link them in the commit message so that it's easier to keep track of the changes later :)

File: circle.yml [hunk 1]
```diff
@@ -3,17 +3,19 @@ dependencies:
     - ~/.pyenv/versions/3.4.3
     - ~/coala-atom/node_modules
   pre:
-    # Install atom dependencies
```
Comment (satwikkansal) [inline circle.yml]: No need to remove this imo. If `libnotify4` is not related to atom-dependencies, just add another comment saying "Install non-atom dependencies" or "Install general dependencies"

File: circle.yml [hunk 2]
```diff
@@ -3,17 +3,19 @@ dependencies:
...
+    # Install  dependencies
+    - sudo apt-get install libnotify4
     - sudo apt-get update
     - sudo apt-get install gvfs-bin
     # Install atom
     - curl -L -o atom.deb https://atom.io/download/deb
     - sudo dpkg --install atom.deb
-
+    - sudo apt-get -f install
```
Comment (satwikkansal) [inline circle.yml]: Please explain in a short comment above the command why do we need to use this "--fix-broken" option

File: circle.yml [hunk 3]
```diff
@@ -3,17 +3,19 @@ dependencies:
...
     # Set python version
     - pyenv local 3.4.3
 
     # Install coala
+    - pip install --upgrade pip
```
Comment (satwikkansal) [inline circle.yml]: Unrelated change? Create a separate issue and commit for it

File: circle.yml (hunk 2 above)
Comment (EverWinter23) [reply circle.yml]: it installs all dependencies for atom, if not already installed.

**Comment by EverWinter23** [comment]
do we really need that comment?

File: circle.yml (hunk 1 above)
Comment (RohanVB) [reply circle.yml]: IMO just `#Install dependencies` as he has it is fine.

File: circle.yml [hunk 4]
```diff
@@ -3,17 +3,24 @@ dependencies:
...
     - sudo dpkg --install atom.deb
-
+    - sudo apt-get -f install
     # Set python version
     - pyenv local 3.4.3
 
+    # install NLTK
+    - pip3 install -U nltk
+    # install NLTK data
+    - python3 -m nltk.downloader punkt maxent_treebank_pos_tagger averaged_perceptron_tagger
```
Comment (RohanVB) [inline circle.yml]: We won't need this after https://github.com/coala/coala-bears/pull/1619; it'll need to be added to `cache_directories` though.

File: circle.yml [hunk 5]
```diff
@@ -3,17 +3,24 @@ dependencies:
...
+
     # Install coala
+    - pip install --upgrade pip
```
Comment (RohanVB) [inline circle.yml]: doesn't the current pip do the job?

File: lib/main.coffee [hunk 6]
```diff
@@ -46,6 +46,16 @@ module.exports =
       lintOnFly: true
       lint: (textEditor) =>
         filePath = textEditor.getPath()
+        # check if project dir contains .coafile
```
Comment (RohanVB) [inline lib/main.coffee]: `check` -> `Check`

Comment (RohanVB) [inline lib/main.coffee]: `check` -> `Check`
`dir` -> `directory`

File: circle.yml (hunk 2 above)
Comment (RohanVB) [reply circle.yml]: I don't think we need to add explanations for flags. :)

Comment (EverWinter23) [reply circle.yml]: thanks

File: lib/main.coffee [hunk 7]
```diff
@@ -46,6 +46,16 @@ module.exports =
...
+        check = helpers.exec(@executable,
+                             ['--find-config'],
+                             {cwd: path.dirname(filePath)}).then (result) ->
+          atom.notifications.addSuccess('Check completed').catch (error) ->
+            atom.notifications.addInfo \
+              'Deactivating coala!',
```
Comment (RohanVB) [inline lib/main.coffee]: I think we need a better notification than coala being deactivated :)

File: circle.yml (hunk 5 above)
Comment (EverWinter23) [reply circle.yml]: no, it it doesn't - check previous builds

File: lib/main.coffee (hunk 7 above)
Comment (EverWinter23) [reply lib/main.coffee]: i'm open to ideas.

Comment (EverWinter23) [reply lib/main.coffee]: other than that, did i resolve the issue?

Comment (RohanVB) [reply lib/main.coffee]: I take it you've tested it locally by running on `atom -d`? 
I'll try it out myself, and write tests for checking it later.
It looks good, if its works then the issue is resolved :)

Comment (EverWinter23) [reply lib/main.coffee]: yeah, i've tested it. It works. I'll make the changes you've requested and update the fork by tomorrow.

**Comment by RohanVB** [comment]
@EverWinter23 You have some extra commits in there (4 total).
Rebase and drop/squash the extras as required please :)

**Comment by EverWinter23** [comment]
Done.

**Comment by RohanVB** [comment]
Not quite :p Still 2 irrelevant commits there.

**Comment by EverWinter23** [comment]
Should I squash them all into one?

**Comment by EverWinter23** [comment]
From coala docs:
What Makes a Good Commit
A good commit is atomic. It should describe one change and not more.

3 commits for 3 changes.

**Comment by RohanVB** [comment]
My bad, I thought you picked up on 2 other commits (didn't notice the change was in the files)

**Comment by EverWinter23** [comment]
So far so good, huh?

File: lib/main.coffee [hunk 8]
```diff
@@ -46,6 +46,16 @@ module.exports =
       lintOnFly: true
       lint: (textEditor) =>
         filePath = textEditor.getPath()
+        # Check if the project directory contains .coafile
+        check = helpers.exec(@executable,
+                             ['--find-config'],
+                             {cwd: path.dirname(filePath)}).then (result) ->
+          atom.notifications.addSuccess('Check completed').catch (error) ->
+            atom.notifications.addInfo \
+              'Missing Dependencies!',
```
Comment (RohanVB) [inline lib/main.coffee]: should we keep it as `coala: Missing ...` ? The package with missing dependencies should be apparent.

Comment (EverWinter23) [reply lib/main.coffee]: Ok, so we have decided on coala: Missing Dependencies?

**Comment by RohanVB** [comment]
You already have the first commit merged here: https://github.com/coala/coala-atom/commit/76a79e1b1c3aef35935b4905f1d17a66ef0b1480
You can drop that one.

File: circle.yml (hunk 4 above)
Comment (RohanVB) [reply circle.yml]: If this is merged before that, 
` install` -> ` Install` in both places

**Comment by EverWinter23** [comment]
Yeah, didn't know about it. But, still have to install NLTK data.

**Comment by EverWinter23** [comment]
Updated with changes.

File: circle.yml [hunk 9]
```diff
@@ -14,6 +14,11 @@ dependencies:
     # Set python version
     - pyenv local 3.4.3
 
+    # Install NLTK
+    - pip3 install -U nltk
+    # Install NLTK data
+    - python3 -m nltk.downloader punkt maxent_treebank_pos_tagger averaged_perceptron_tagger
+
```
Comment (RohanVB) [inline circle.yml]: We don't need this commit anymore, nltk data will be loaded from cache_directories :)

Comment (EverWinter23) [reply circle.yml]: I'll rebase the fork.

**Comment by EverWinter23** [comment]
how to apply the above patch?

**Comment by satwikkansal** [comment]
> how to apply the above patch?

@EverWinter23 just run coala locally ;)

**Comment by RohanVB** [comment]
@satwikkansal check out this issue: https://github.com/coala/coala-bears/issues/1542
He'll have to fix it manually or use `git apply`. Running coala would delete the contents of the .md file

**Comment by satwikkansal** [comment]
@RohanVB Thanks. I don't think there's need to apply the patch as the PR doesn't include any modifications to `README.md`. Looks like this has to wait until the issues with Markdown Bear gets fixed, or we may have to merge it as is. IIRC, there have been a few PRs by jayvdb to fix the incompatibility issues with `remark`.

**Comment by RohanVB** [comment]
> Looks like this has to wait until the issues with Markdown Bear gets fixed

The better option is applying the patch manually :) 

>  merge it as is

Right now he has merge conflicts on the .md file

File: README.md [hunk 10]
```diff
@@ -17,8 +17,8 @@ editor.
 
 # AUTHORS
 
-*   Abdeali J Kothari - <mailto:abdealikothari@gmail.com>
-*   Lasse Schuirmann - <mailto:lasse.schuirmann@gmail.com>
+Abdeali J Kothari - <mailto:abdealikothari@gmail.com>
+Lasse Schuirmann - <mailto:lasse.schuirmann@gmail.com>
```
Comment (RohanVB) [inline README.md]: It should look like 
```
- Abdeali J Kothari - <mailto:abdealikothari@gmail.com>
- Lasse Schuirmann - <mailto:lasse.schuirmann@gmail.com>
```

File: lib/main.coffee [hunk 11]
```diff
@@ -46,6 +46,16 @@ module.exports =
...
+        # Check if the project directory contains .coafile
+        check = helpers.exec(@executable,
+                             ['--find-config'],
+                             {cwd: path.dirname(filePath)}).then (result) ->
+          atom.notifications.addSuccess('Check completed').catch (error) ->
+            atom.notifications.addInfo \
+              'coala: Missing Dependencies!',
+              'detail': 'Project directory does not contain .coafile.\n
+                         Please provide a .coafile for using coala.\n
+                         Make sure coala-bears meet all dependencies.'
```
Comment (RohanVB) [inline lib/main.coffee:58]: Are you sure this works as intended? When I test with removing the `.coafile` and/or `coala-bears`, the only error that's raised is from the base `linter`, which flags: `See Console for more info. (Open View -> Developer -> Toogle Developer Tools)`

File: README.md [hunk 12]
```diff
@@ -17,8 +17,8 @@ editor.
 
 # AUTHORS
 
-*   Abdeali J Kothari - <mailto:abdealikothari@gmail.com>
-*   Lasse Schuirmann - <mailto:lasse.schuirmann@gmail.com>
+- Abdeali J Kothari - <mailto:abdealikothari@gmail.com>
```
Comment (supergr35) [inline README.md]: unrelated change? apparently it doesn't fix the gitmate error

Comment (sils) [reply README.md]: yep the problem here is that it's broken in one commit and fixed in the other, gitmate doesn't want you to break it in the first place basically

Comment (sils) [reply README.md]: uh interesting... maybe just switch out the commits?

Comment (EverWinter23) [reply README.md]: tried that, didn't touch the README.md, still got the error.

Comment (sils) [reply README.md]: sorry for not responding for so long, let's just ignore gitmate on this one then

File: lib/main.coffee (hunk 11 above)
Comment (sils) [reply lib/main.coffee:58]: same for me.

**Comment by sils** [comment]
I rebased it for you and made gitmate and tests pass (force push!) but it doesn't seem to work :/

**Comment by EverWinter23** [comment]
Thanks
